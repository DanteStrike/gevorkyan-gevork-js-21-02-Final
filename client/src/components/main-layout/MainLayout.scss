@import "src/assets/mixins";

.main-layout {
  display: grid;
  grid-template-areas: "header" "content" "footer";
  grid-template-rows: auto 1fr auto;
  grid-template-columns: 1fr;
  grid-gap: 15px;
  height: 100%;

  &__header {
    grid-area: header;
  }

  &__content {
    @include container;
    display: flex;
    grid-area: content;
    align-items: center;
    justify-content: center;
  }

  &__footer {
    grid-area: footer;
  }
}
