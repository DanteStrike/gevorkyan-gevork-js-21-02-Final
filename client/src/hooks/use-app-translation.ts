import {useTranslation} from 'react-i18next';

const useAppTranslation = (nestSpace?: string) => useTranslation(`translation`, {keyPrefix: nestSpace});

export default useAppTranslation;
