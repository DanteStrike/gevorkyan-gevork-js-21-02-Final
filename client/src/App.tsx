import React, {useEffect, useState} from 'react';
import './App.scss';
import {message} from 'antd';
import useAppSelector from './hooks/use-app-selector';
import {authOperations, authSelectors} from './store/auth';
import Loading from './components/loading/Loading';
import MainLayout from './components/main-layout/MainLayout';
import {authStorageKey} from './store/auth/types';
import useAppDispatch from './hooks/use-app-dispatch';

const enum AppState {
  INIT = `init`,
  WAITING = `waiting`,
  READY = `ready`,
}

function App() {
  const [appSetup, setAppSetup] = useState<AppState>(AppState.INIT);
  const dispatch = useAppDispatch();

  if (appSetup === AppState.INIT) {
    const authStorageID = localStorage.getItem(authStorageKey) || ``;
    if (authStorageID !== ``) {
      dispatch(authOperations.login(authStorageID));
      setAppSetup(AppState.WAITING);
    } else {
      setAppSetup(AppState.READY);
    }
  }

  const isAuthWait = useAppSelector(authSelectors.getIsAuthWait);
  const isAuthError = useAppSelector(authSelectors.getIsError);

  useEffect(() => {
    if (appSetup === AppState.WAITING && isAuthError && isAuthWait) {
      message.error(`Ошибка авто авторизации`);
      setAppSetup(AppState.READY);
    }

    if (appSetup === AppState.WAITING && !isAuthError && !isAuthWait) {
      setAppSetup(AppState.READY);
    }
  }, [appSetup, isAuthError, isAuthWait]);

  if (appSetup !== AppState.READY) {
    return <Loading isLoading />;
  }

  return <MainLayout />;
}

export default App;
