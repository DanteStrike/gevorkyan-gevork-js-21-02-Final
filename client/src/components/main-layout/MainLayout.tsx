import React from 'react';
import './MainLayout.scss';
import Header from '../header/Header';
import Routes from '../routes/Routes';
import Footer from '../footer/Footer';

function MainLayout() {
  return (
    <div className="main-layout">
      <div className="main-layout__header">
        <Header />
      </div>
      <div className="main-layout__content">
        <Routes />
      </div>
      <div className="main-layout__footer">
        <Footer />
      </div>
    </div>
  );
}

export default React.memo(MainLayout);
