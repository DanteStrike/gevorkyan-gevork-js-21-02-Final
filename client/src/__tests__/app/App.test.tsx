import {mount, shallow} from 'enzyme';
import React from 'react';
import App from '../../App';
import * as useAppSelector from '../../hooks/use-app-selector';
import * as useAppDispatch from '../../hooks/use-app-dispatch';
import MainLayout from '../../components/main-layout/MainLayout';

describe(`App should render correctly`, () => {
  it(`mode ready snapshot`, () => {
    jest.spyOn(useAppSelector, `default`).mockReturnValue(false);
    jest.spyOn(useAppDispatch, `default`).mockReturnValue(jest.fn());

    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it(`mode waiting snapshot`, () => {
    jest
      .spyOn(useAppSelector, `default`)
      .mockReturnValueOnce(true)
      .mockReturnValueOnce(false)
      .mockReturnValueOnce(true)
      .mockReturnValueOnce(false);
    jest.spyOn(useAppDispatch, `default`).mockReturnValue(jest.fn());
    localStorage.setItem(`app-auth-id`, `anyID`);

    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`App should work correctly`, () => {
  it(`success loading snapshot`, () => {
    jest.spyOn(MainLayout, `type`).mockImplementation(() => <div id="main-content" />);
    jest.spyOn(useAppSelector, `default`).mockReturnValue(false);
    jest.spyOn(useAppDispatch, `default`).mockReturnValue(jest.fn());
    localStorage.setItem(`app-auth-id`, `anyID`);

    const wrapper = mount(<App />);
    expect(wrapper.find(`#main-content`)).toHaveLength(1);
  });

  it(`error loading snapshot`, () => {
    jest.spyOn(MainLayout, `type`).mockImplementation(() => <div id="main-content" />);
    jest.spyOn(useAppSelector, `default`).mockReturnValue(true);
    jest.spyOn(useAppDispatch, `default`).mockReturnValue(jest.fn());
    localStorage.setItem(`app-auth-id`, `anyID`);

    const wrapper = mount(<App />);
    expect(wrapper.find(`#main-content`)).toHaveLength(1);
  });
});
