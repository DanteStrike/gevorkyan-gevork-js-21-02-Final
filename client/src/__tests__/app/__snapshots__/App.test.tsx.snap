// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly mode ready snapshot 1`] = `<Memo(MainLayout) />`;

exports[`App should render correctly mode waiting snapshot 1`] = `
<Memo(Loading)
  isLoading={true}
/>
`;
